File: lib/core/models/shopping_item_model.dart
import 'dart:math';

import 'package:faker/faker.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import 'package:rcp/core/models/shopping_list_model.dart';

part 'shopping_item_model.freezed.dart';
part 'shopping_item_model.g.dart';

@freezed
class ShoppingItem with _$ShoppingItem {
  const ShoppingItem._();
  @JsonSerializable(explicitToJson: true)
  factory ShoppingItem({
    required String id,
    required String shoppingListId,
    required String name,
    String? quantity,
    UserProfile2? purchasedBy,
    DateTime? purchasedAt,
    required DateTime createdAt,
    required UserProfile2 createdBy,
  }) = _ShoppingItem;

  factory ShoppingItem.fromJson(Map<String, dynamic> json) =>
      _$ShoppingItemFromJson(json);

  bool get isPurchased => purchasedBy != null && purchasedAt != null;
}

List<ShoppingItem> generateFakeShoppingItemData(int count) {
  List<ShoppingItem> fakeData = [];
  Faker faker = Faker();
  Random random = Random();

  for (int i = 0; i < count; i++) {
    String id = faker.guid.guid();
    String shoppingListId = faker.guid.guid();
    String name = faker.food.cuisine();

    DateTime createdAt = faker.date.dateTime();
    UserProfile2 createdBy = generateFakeUserProfileData();
    DateTime updatedAt = faker.date.dateTime();
    UserProfile2 updatedBy = generateFakeUserProfileData();
    bool isPurchased = random.nextBool();

    fakeData.add(ShoppingItem(
      id: id,
      shoppingListId: shoppingListId,
      name: name,
      createdAt: createdAt,
      createdBy: createdBy,
      quantity: '2 box',
      purchasedAt: isPurchased ? updatedAt : null,
      purchasedBy: isPurchased ? updatedBy : null,
    ));
  }

  return fakeData;
}

File: supabase/functions/shared/models/shopping_item_model.ts
import { faker } from 'npm:@faker-js/faker';
import { UserProfile } from './user_profile_model.ts';
import { Database } from '../../types.ts';

export type ShoppingItemRow =
	Database['public']['Tables']['shopping_items']['Row'];

export const shoppingItemTable = 'shopping_items';

export const shoppingItemSelect = `
id: id,
shoppingListId: shopping_list_id,
name: name,
quantity: quantity,
purchasedAt: purchased_at,
purchasedBy: purchased_by (
	id: id,
	userId: user_id,
	username: username,
	fullName: full_name,
	avatarUrl: avatar_url,
	createdAt: created_at,
	updatedAt: updated_at
),
updatedAt: updated_at,
updatedBy: updated_by (
	id: id,
	userId: user_id,
	username: username,
	fullName: full_name,
	avatarUrl: avatar_url,
	createdAt: created_at,
	updatedAt: updated_at
),
createdAt: created_at,
createdBy: created_by (
	id: id,
	userId: user_id,
	username: username,
	fullName: full_name,
	avatarUrl: avatar_url,
	createdAt: created_at,
	updatedAt: updated_at
)
`;

interface ShoppingItem {
	id: string;
	shoppingListId: string;
	name: string;
	quantity?: string | null;
	purchasedAt?: Date | null;
	purchasedBy?: UserProfile | null;
	createdAt: Date;
	createdBy: UserProfile;
	updatedAt: Date;
	updatedBy: UserProfile;
}

export type { ShoppingItem };
